### PR TITLE
feat: Activate Version with example test

### LINKED ISSUE #72: Support for activate version

### LINKED ISSUE #88: Add sdc_metadata to records
Looks to be implemented in the SDK but I"m not certain how to turn this on in the target

### PR DESCRIPTION
**Pull request by visch**
Closes #72 
Closes #88

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,20 +14,22 @@ Built with the [Meltano SDK](https://sdk.meltano.com) for Singer Taps and Target
...
+| dialect+driver        |  False   | postgresql+psycopg2 | Dialect+driver see https://docs.sqlalchemy.org/en/20/core/engines.html. Generally just leave this alone. Note if sqlalchemy_url is set this will be ignored.                                                                                                           |
+| default_target_schema |  False   |        None         | Postgres schema to send data to, example: tap-clickup                                                                                                                                                                                                                  |
+| hard_delete           |  False   |          0          | When activate version is sent from a tap this specefies if we should delete the records that don't match, or mark them with a date in the `_sdc_deleted_at` column.                                                                                                    |
```
Comment (visch) [inline README.md:27]: Boolean should be generated as True / False here I think? I'll leave for now

File: target_postgres/sinks.py [hunk 2]
```diff
@@ -252,3 +255,62 @@ def schema_name(self) -> Optional[str]:
 
         # Schema name not detected.
         return None
+
+    def activate_version(self, new_version: int) -> None:
```
Comment (visch) [inline target_postgres/sinks.py:259]: I think we could take this function and just drop it into the SDK as it is here, this just has a few tweaks that are needed to make this work in a more generic way

File: target_postgres/sinks.py [hunk 3]
```diff
@@ -252,3 +255,62 @@ def schema_name(self) -> Optional[str]:
...
+                self.full_table_name,
+                self.version_column_name,
+                sql_type=integer_type,  # Different from SingerSDK as we need to handle types the same as SCHEMA messsages
+            )
+
+        self.logger.info("Hard delete: %s", self.config.get("hard_delete"))
+        if self.config["hard_delete"] is True:
+            self.connection.execute(
+                f"DELETE FROM {self.full_table_name} "
+                f"WHERE {self.version_column_name} <= {new_version} OR {self.version_column_name} IS NULL"
```
Comment (visch) [inline target_postgres/sinks.py:292]: I'm a little worried about what would happen if someone ran this in production with `add_record_metadata` set to True for a while, then suddenly they change it to `False`. 

I think we'd end up dropping the entire table, because we'd add all the new records, then none of them would have a `new_version` set they'd all be NULL. Then we'd drop all the NULLs.  :O 

Removing the IS NULL check would work here but I think the better thing to do would be a check to be sure metadata_columns is enabled as I think the IS NULL check is very valid here as in normal operation you want activate_version to be keeping your data very "clean" and in the same shape as the source. 

I should also write a test for this case as it's potentially scary (only place in our code that we drop tables so being cautious is good)

File: target_postgres/sinks.py [hunk 4]
```diff
@@ -252,3 +255,62 @@ def schema_name(self) -> Optional[str]:
...
+            self.connector.prepare_column(
+                self.full_table_name,
+                self.version_column_name,
+                sql_type=integer_type,  # Different from SingerSDK as we need to handle types the same as SCHEMA messsages
+            )
+
+        self.logger.info("Hard delete: %s", self.config.get("hard_delete"))
+        if self.config["hard_delete"] is True:
+            self.connection.execute(
+                f"DELETE FROM {self.full_table_name} "
```
Comment (visch) [inline target_postgres/sinks.py:291]: If you look at `activate_version_hard.singer` There is an `ACTIVATE_VERSION` message before the records, and an `ACTIVATE_VERSION` message after the records. I believe right now when we get the first `ACTIVATE_VERSION` message we will end up dropping the table. 

While I think two `ACTIVATE_VERSION` messages is overkill we should support the `transferwise` ACTIVATE_VERSION style without being destructive. Need to test this

Comment (visch) [reply target_postgres/sinks.py:294]: Have a failing test added now for this. We need to do something here, maybe on the first ActivateVersion message we don't delete?

File: target_postgres/sinks.py (hunk 3 above)
Comment (visch) [reply target_postgres/sinks.py:295]: If we fix the other issue https://github.com/MeltanoLabs/target-postgres/pull/89/files#r1084618420 I believe this one gets solved as well.

File: target_postgres/tests/test_standard_target.py [hunk 5]
```diff
@@ -232,3 +244,94 @@ def test_new_array_column(postgres_target):
...
+        result = connection.execute("SELECT * FROM test_activate_version_soft")
+        assert result.rowcount == 9
+
+        result = connection.execute(
+            "SELECT * FROM test_activate_version_soft where _sdc_deleted_at is NOT NULL"
+        )
+        assert result.rowcount == 2
+
+
+def test_activate_version_doesnt_delete_before_populating(postgres_config):
```
Comment (visch) [inline target_postgres/tests/test_standard_target.py:315]: This fails right now re https://github.com/MeltanoLabs/target-postgres/pull/89/files#r1084618420

Comment (visch) [reply target_postgres/tests/test_standard_target.py:315]: This may be desired behavior, https://github.com/transferwise/pipelinewise-tap-postgres/blob/b78ba51f3ff239c8de762f6eaa5406f53ec7b964/tap_postgres/sync_strategies/full_table.py 

I think the idea is if it's the first run you want to drop the table first. Hmmmm seems like an odd implementation but it doesn't look like a spec issue honestly.

File: target_postgres/sinks.py (hunk 4 above)
Comment (edgarrmondragon) [reply target_postgres/sinks.py:294]: Way back I remember seeing intermediate tables for specific versions, e.g. `users_1674486431563` so maybe the first message could be a signal to create such a temp table?

Comment (visch) [reply target_postgres/sinks.py:294]: @edgarrmondragon hmm I think the first message in the previous example is actually on purpose by the tap-postgres transfer wise implementation. I have no idea why they'd want to do that but I think they want to truncate the table before they push any data.

Ie my concern here isn't valid, I think anytime an activate version message is sent at should truncate the table (based on version)

File: target_postgres/target.py [hunk 6]
```diff
@@ -71,7 +71,7 @@ def __init__(
             th.StringType,
             description=(
                 "User name used to authenticate. "
-                + "Note if sqlalchemy_url is set this will be ignored.",
```
[the remark this answers is not in the document]
Comment (edgarrmondragon) [reply target_postgres/target.py:74]: This would've been caught by `mypy` so it's probably worth adding that to the pre-commit config.

**Review by edgarrmondragon — COMMENTED** [review]
This is looking great so far @visch!

File: target_postgres/sinks.py (hunk 4 above)
Comment (visch) [reply target_postgres/sinks.py:294]: As I thought about this more.

We do know if a tap is outputting an `ACTIVATE_VERSION` message before sending any records. We could just ignore the first one or throw an error/warning when this happens. In my opinion this is just wrong but maybe we should treat it well as that's what `pipelinewise` does in their taps. If we started adding functionality around this behavior then I think we should expand the spec for `ACTIVATE_VERSION`

**Review by edgarrmondragon — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (aaronsteers) [reply README.md:27]: Agreed. Very strange that it shows defaults of `0`. Agreed though we don't need to block on this.

File: README.md [hunk 7]
```diff
@@ -14,20 +14,22 @@ Built with the [Meltano SDK](https://sdk.meltano.com) for Singer Taps and Target
...
+| password              |  False   |        None         | Password used to authenticate. Note if sqlalchemy_url is set this will be ignored.                                                                                                                                                                                     |
+| database              |  False   |        None         | Database name. Note if sqlalchemy_url is set this will be ignored.                                                                                                                                                                                                     |
+| sqlalchemy_url        |  False   |        None         | SQLAlchemy connection string. This will override using host, user, password, port,dialect. Note that you must esacpe password specialcharacters properly seehttps://docs.sqlalchemy.org/en/20/core/engines.html#escaping-special-characters-such-as-signs-in-passwords |
```
Comment (aaronsteers) [inline README.md:24]:
```suggestion
| sqlalchemy_url        |  False   |        None         | SQLAlchemy connection string. This will override using host, user, password, port,dialect. Note that you must escape password special characters properly. See https://docs.sqlalchemy.org/en/20/core/engines.html#escaping-special-characters-such-as-signs-in-passwords |
```

File: target_postgres/sinks.py (hunk 4 above)
Comment (aaronsteers) [reply target_postgres/sinks.py:294]: > There is an ACTIVATE_VERSION message before the records, and an ACTIVATE_VERSION message after the records.

TL;DR - I don't think we need to do anything special in targets to handle an initial and then final `ACTIVATE_VERSION` message. We can handle both as we would either - which is to delete (or flag for deletion) any records which the version makes obsolete.

My understanding is that the initial `ACTIVATE_VERSION` message _only_ is emitted if there is no prior `STATE` bookmark for this particular stream, meaning we think we are (probably) syncing the stream for the first time. The rationale that I heard for sending twice is that there's a _potential_ optimization of making records immediately available as they land in the destination. I think the optimization benefit is dubious and I don't know if we would carry over the same in SDK-based taps. Either way, I think that's safely outside the scope of things we need to solve in this PR. 😄 

Here's a sample tap implementation where you can see the initial `ACTIVATE_VERSION` message is conditional based on the `if first_run:` check.

https://github.com/transferwise/pipelinewise-tap-postgres/blob/a97d7d87493ad5b60b7d016fecf5100f91a5c4ec/tap_postgres/sync_strategies/full_table.py#L40-L41

**Review by aaronsteers — APPROVED** [review]
